### PR TITLE
DOC: improve docs on iteration

### LINKED ISSUE #7194: DOC: df.iterrows() returns view for homogenous columns, and copy for heterogenous / perf warning
perf discussion: http://stackoverflow.com/questions/24870953/does-iterrows-have-performance-issues/24871316#24871316

```
import pandas as pd
df = pd.DataFrame({'a':range(5),'b':range(5)})
print df

for _, row in df.iterrows():
    row.a += 1
print df

df['c'] = 'what'

for _, row in df.iterrows():
    row.b += 1
print df

   a  b
0  0  0
1  1  1
2  2  2
3  3  3
4  4  4

[5 rows x 2 columns]
   a  b
0  1  0
1  2  1
2  3  2
3  4  3
4  5  4

[5 rows x 2 columns]
   a  b     c
0  1  0  what
1  2  1  what
2  3  2  what
3  4  3  what
4  5  4  what

[5 rows x 3 columns]
```

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
As this comes up once in a while, I tried to improve/clarify the docs somewhat.

Further suggestions to add or make it clearer are welcome!

xref #10334 

closes #7194

### CODE REVIEW

File: doc/source/basics.rst [hunk 1]
```diff
@@ -1154,21 +1154,40 @@ any of its three axes.
...
+       ...:     print(col)
+       ...:
+
+Pandas objects also have the dict-like :meth:`~DataFrame.iteritems` method to
+iterate over the (key, value) pairs.
+
+To iterate over the rows of a DataFrame, you can use the following methods:
+
+* :meth:`~DataFrame.iterrows`: Iterate over the rows of a DataFrame as (index, Series) pairs.
+* :meth:`~DataFrame.itertuples`: Iterate over the rows of a DataFrame as tuples of the values.
```
Comment (shoyer) [inline doc/source/basics.rst]: I would note that this is faster than `iterrows` is almost all cases

**Comment by jorisvandenbossche** [comment]
@shoyer thanks, I forgot to note that, added now!

File: doc/source/basics.rst [hunk 2]
```diff
@@ -1154,21 +1154,46 @@ any of its three axes.
...
+  performance implications.
+* :meth:`~DataFrame.itertuples`: Iterate over the rows of a DataFrame as tuples of the values.
+  This is a lot faster as :meth:`~DataFrame.iterrows`, and is in most cases preferable to
+  use to iterate over the values of a DataFrame.
+
+.. warning::
+
+  In many cases, iterating manually over the rows is not needed. Iterating with
+  pandas objects is generally slow, and in most cases vectorized solutions
+  exist.
```
Comment (jreback) [inline doc/source/basics.rst]: you also need a big warning that you cannot SET columns and expect it to work (as its operating on a copy). This is also a big source of confusion. 

Further you can mention ``.apply``.

I would highlite the performance issues WAY more here. Encourage vectorization first.

File: doc/source/basics.rst [hunk 3]
```diff
@@ -1189,22 +1214,46 @@ For example:
...
+      df_orig.dtypes
+      row = next(df_orig.iterrows())[1]
+      row
+
+   All values in ``row``, returned as a Series, are now upcasted
+   to floats, also the original integer value in column `x`:
+
+   .. ipython:: python
+
+      print(row['int'].dtype)
```
Comment (jreback) [inline doc/source/basics.rst]: you don't need the print

Comment (jorisvandenbossche) [inline doc/source/basics.rst]: yep, was just copied from before

**Comment by jorisvandenbossche** [comment]
@jreback @shoyer I added some further warnings / alternatives (also about setting in a loop, so this should now also address #7194). You can have a new look.

File: doc/source/basics.rst [hunk 4]
```diff
@@ -1154,21 +1154,76 @@ any of its three axes.
...
+  * Look for *vectorized* solution: many operations can be performed using
+    built-in methods or numpy functions, (boolean) indexing.
+
+  * When you have a function that cannot work on the full DataFrame/Series
+    at once, it is better to use :meth:`~DataFrame.apply` instead of iterating
+    over the values. See the docs on :ref:`function application <basics.apply>`.
+
+  * If you need to do iterative manipulations on the values but performance is
+    important, consider writing the inner loop using e.g. cython or numba.
+    See the :ref:`enhancing performance <_enhancingperf>` section for some
```
Comment (jreback) [inline doc/source/basics.rst]: weird why this is ``_enhancingperf``, rather than ``enhancingperf`` like everything else.....

**Comment by jreback** [comment]
looks great! (minor comment on the reference to enhancingperf)

File: doc/source/basics.rst (hunk 4 above)
Comment (jorisvandenbossche) [inline doc/source/basics.rst]: good catch!

**Pull request merged**